[WIP] feat(overlay): Structured Logging support

Shubhdeep12 opened the pull request:
<!--
Tick these boxes if they're applicable to your PR.
- Changesets are only required for PRs to Spotlight library packages (e.g. @spotlightjs/overlay). Not for the website/docs or demo app contributions.
- Typo correction or small bugfix PRs don't require an issue. If you're making a bigger change, please open an issue first.
-->

Before opening this PR:

- [ ] I added a [Changeset Entry](https://spotlightjs.com/contribute/changesets/) with `pnpm changeset:add`
- [x] I referenced issues that this PR addresses
#756 


- [x] To accept Log Envelopes
- [x] To list all logs
- [ ] To display details of a single log and a link back to the trace.
- [ ] Log type envelopes should be displayed in the Envelopes tab.

BYK commented on .changeset/bitter-turtles-brake.md:
```diff
@@ -0,0 +1,5 @@
+---
+'@spotlightjs/overlay': major
```
Why major, do we have a breaking change somewhere?

BYK commented on packages/overlay/src/integrations/console/console-tab.tsx:
```diff
@@ -1,3 +1,4 @@
+import CardList from '~/components/CardList';
```
With your patch, I think we should just get rid of this extension 😅

BYK commented on packages/overlay/src/integrations/sentry/store/slices/eventsSlice.ts:
```diff
@@ -39,11 +40,23 @@ export const createEventsSlice: StateCreator<SentryStore, [], [], EventsSliceSta
+      if (!event.items) {
+        event.items = [];
+      }
+      event.items.forEach((logItem: SentryLogEventItem) => {
```
Prefer a for-of loop please

Shubhdeep12 replied:
sure.

BYK commented on packages/overlay/src/integrations/sentry/utils/duration.ts:
```diff
@@ -35,3 +35,18 @@ export function getFormattedDuration(duration: number): string {
   return getFormattedDuration(span.timestamp - span.start_timestamp);
 }
+
+export function formatTimestamp(timestamp: number) {
```
Can we just stick to ISO-8601 format instead of US formatting? Pretty sure the rest of the world will thank us later on along with some US folks themselves.

betegon commented on packages/overlay/src/integrations/sentry/types.ts:
```diff
@@ -201,7 +201,28 @@ export type SentryProfileV1Event = CommonEventAttrs & {
 };
 
-export type SentryEvent = SentryErrorEvent | SentryTransactionEvent | SentryProfileV1Event;
+export type SentryLogEventItem = {
```
Good catch! We actually have `SerializedLog` [here](https://github.com/getsentry/sentry-javascript/blob/a79ca7eb30794d0aa40e3d0a47853ad5fc10ef64/packages/core/src/types-hoist/log.ts#L39-L69). I think that is what we should use here @Shubhdeep12 .

Shubhdeep12 replied:
Let me use the type SerializedLog, but I’ll extend it and add a `log_id` property, since we need it as a unique ID attached to each log.

Shubhdeep12 commented on packages/overlay/src/integrations/sentry/components/insights/index.tsx:
```diff
@@ -27,6 +28,7 @@ export default function InsightsTabDetails() {
     createTab('webvitals', 'Web Vitals'),
     createTab('resources', 'Resources'),
     createTab('profiles', 'Profiles'),
+    createTab('logs', 'Logs'),
```
sure lets move this on top

Shubhdeep12 commented on packages/overlay/src/integrations/sentry/store/slices/eventsSlice.ts:
```diff
@@ -39,11 +40,23 @@ export const createEventsSlice: StateCreator<SentryStore, [], [], EventsSliceSta
     }
 
+    if (isLogEvent(event)) {
+      if (!event.items) {
```
🙈

Shubhdeep12 commented on packages/overlay/src/integrations/sentry/data/useSentryLogs.tsx:
```diff
@@ -0,0 +1,25 @@
+  const getEvents = useSentryStore(state => state.getEvents);
+  const getEventsByTrace = useSentryStore(state => state.getEventsByTrace);
+
+  const allEvents = traceId ? getEventsByTrace(traceId) : getEvents();
```
This will not work @BYK @betegon 

Because in Log Events, the trace_id is present in log items, as different trace_ids can be attached to log items within a single Log Event.
ref: https://discord.com/channels/621778831602221064/1376064165076271195/1376064165076271195


fixing this.

Shubhdeep12 replied:
fixed

Shubhdeep12 commented:
hey @BYK @betegon 

following changes are added in the PR:
1. Fixed the useSentryLogs hook to return the logs for a specific trace.
2. Added a Logs tab under each trace as well. Taking sentry.io dashboard as reference.
![image](https://github.com/user-attachments/assets/516835e1-ee1f-4586-b9d4-eb3dd3410921)
![image](https://github.com/user-attachments/assets/057f86a5-ab66-4182-b1a8-c4f2d7159acb)

BYK commented:
@betegon
>@BYK wouldn't this PR also close https://github.com/getsentry/spotlight/issues/643 ?

Not really unfortunately. The SDKs own logging (enabled via `debug:true`) are exempt from structured logging. We need them to capture these too when `spotlight:true` and `debug:true` I guess. We may also want to look into "client reports" where the SDKs report when an event is discarded due to size etc.

BYK reviewed (changes requested):
Didn't go in-depth but looks good overall. Can use an updated screenshot or video capture for the PR description as we moved logs to top level now.

The `itemTypesList` thing I mentioned is confusing and we should set the changelog entry to say `minor` as I suggested.

I'll do a final pass before merging unless @betegon beats me to it.

Shubhdeep12 commented on packages/overlay/src/integrations/sentry/components/insights/envelopes/EnvelopeList.tsx:
```diff
@@ -32,6 +32,8 @@ export default function EnvelopeList({ showAll }: { showAll: boolean }) {
               const { trace_id } = (header?.trace as { trace_id?: string }) || {};
               const envelopeItems = envelope[1] || [];
               const itemTypes = new Set();
+              const itemTypesList = Array.from(itemTypes).join(',');
```
fixed

Shubhdeep12 commented on packages/overlay/src/integrations/sentry/store/slices/eventsSlice.ts:
```diff
@@ -39,11 +39,20 @@ export const createEventsSlice: StateCreator<SentryStore, [], [], EventsSliceSta
 
+    if (isLogEvent(event) && event.items?.length) {
+      for (const logItem of event.items) {
+        logItem.timestamp = toTimestamp(logItem.timestamp);
```
I think its fine here for log.
https://docs.sentry.io/platforms/javascript/logs/#integrations:~:text=timestamp%3A%20(number)%20The%20timestamp%20of%20the%20log.

timestamp will always be there

BYK commented on packages/overlay/src/integrations/sentry/store/slices/eventsSlice.ts:
```diff
@@ -39,11 +39,20 @@ export const createEventsSlice: StateCreator<SentryStore, [], [], EventsSliceSta
+    if (isLogEvent(event) && event.items?.length) {
+      for (const logItem of event.items) {
+        logItem.timestamp = toTimestamp(logItem.timestamp);
+        logItem.log_id = generateUuidv4();
```
Nah, I think we're good for now.

betegon commented on packages/overlay/src/integrations/sentry/store/types.ts:
```diff
@@ -14,7 +22,7 @@ export type Subscription = OnlineSubscription | EventSubscription | TraceSubscri
 export interface EventsSliceState {
   events: SentryEvent[];
-  eventIds: Set<string>;
+  eventsById: Map<string, SentryEvent>;
```
this makes much more sense. from what I've heard you offline, we are going to go with maps for every id related event, right?

BYK replied:
yup, 100%

betegon commented on packages/overlay/src/integrations/sentry/store/slices/eventsSlice.ts:
```diff
@@ -28,22 +29,48 @@ export const createEventsSlice: StateCreator<SentryStore, [], [], EventsSliceSta
+        }
+        logItem.sdk = logItem.attributes?.['sentry.sdk.name']?.value as string;
+        logItem.timestamp = toTimestamp(logItem.timestamp);
+        logItem.id = generateUuidv4();
```
weren't we going to add a prefix to ids?

BYK replied:
Yes but in a follow up :)

betegon commented on packages/overlay/src/integrations/sentry/data/useSentryLogs.tsx:
```diff
@@ -0,0 +1,27 @@
+    logsByTraceId: state.logsByTraceId,
+  }));
+
+  const allLogs = Array.from(traceId ? (logsByTraceId.get(traceId) ?? []) : logsById.values());
+  const localLogs = allLogs.filter(item => item.trace_id && isLocalTrace(item.trace_id));
```
shouldn't we memoize these?

```typescript
  const allLogs = useMemo(
    () => Array.from(traceId ? (logsByTraceId.get(traceId) ?? []) : logsById.values()),
    [traceId, logsById, logsByTraceId]
  );
  const localLogs = useMemo(
    () => allLogs.filter(item => item.trace_id && isLocalTrace(item.trace_id)),
    [allLogs]
  );
```

BYK replied:
I think this might be done downstream? I'm also not very clear on `useMemo()`. Maybe try in a follow up?

betegon reviewed (changes requested):
Looks good. It’s a bit weird that the logs don’t have IDs, but I guess there’s not much we can do about that.

BYK reviewed (approved):
let me approve my own changes 🤣
